Tirpc

skoudmar opened the pull request.

CendioOssman committed:
Switch to using libtirpc

The old stuff has been removed from glibc so we need to start relying on
an independent RPC library.

This also converts everything to IPv6, as that is required by libtirpc.

Fixes from Martin Škoudlil <skoudmar@fel.cvut.cz>

CendioOssman committed:
Better parsing errors

There are a lot of things that can go wrong, so let's try to be more
clear about what we don't like in the exports file.

CendioOssman committed:
Handle socket() errors

Otherwise we'll just get a confusing "bad file descriptor" when we try
to call bind().

CendioOssman committed:
Support systems with IPv6 disabled

Fixes from Martin Škoudlil <skoudmar@fel.cvut.cz>

wentasah commented on daemon.c:
```diff
@@ -847,6 +847,11 @@ static SVCXPRT *create_tcp_transport(unsigned int port)
 	    fprintf(stderr, "Couldn't bind to tcp port %d\n", port);
 	    exit(1);
 	}
+	if (listen(sock, SOMAXCONN)) {
```
Je mi trochu divné, že listen je potřeba jen v `create_tcp_transport()` a ne jinde. Pokud to testujete jen s TCP a UDP neřešíte, tak je potřeba to napsat do commit message.

skoudmar replied:
Protokol UDP nepoužívá listen. Testoval jsem i UDP a funguje.

wentasah commented on daemon.c:
```diff
@@ -13,6 +13,8 @@
 #include <rpc/rpc.h>
 #include <rpc/pmap_clnt.h>
+#include <netconfig.h>
+
```
Asi zbytečný newline.

wentasah commented on daemon.c:
```diff
@@ -673,6 +675,25 @@ static void register_nfs_service(SVCXPRT * udptransp, SVCXPRT * tcptransp)
+	    struct netconfig* nconf;
+	    /* Returns a pointer to nconf corresponding to NETCONF */
+	    if ((nconf = getnetconfigent("udp6")) ==
+		(struct netconfig *)NULL) {
```
NULL by nemělo být potřeba přetypovávat.

wentasah commented on daemon.c:
```diff
@@ -673,6 +675,25 @@ static void register_nfs_service(SVCXPRT * udptransp, SVCXPRT * tcptransp)
+		(struct netconfig *)NULL) {
+	    fprintf(stderr, "cannot get netconfig entry for udp6\n");
+	    daemon_exit(0);
+	    }	
```
Tady je nějaký zbytečný "trailing whitespace". Problémy se zarovnáním (zatím) neřeším.

wentasah commented on daemon.c:
```diff
@@ -673,6 +675,25 @@ static void register_nfs_service(SVCXPRT * udptransp, SVCXPRT * tcptransp)
 	}
+
+	if (opt_portmapper) {
+	    // FIXME: do not register if using only ipv4
```
Koukám, že to FIXME řešíte v nějakém dalším commitu. Buďto napište do commit message, že to vyřešíte v dalším comitu, ale lepší by bylo, když to vyřešíte rovnou v tomto commitu.

wentasah commented on daemon.c:
```diff
@@ -251,6 +252,7 @@ static void parse_options(int argc, char **argv)
 		printf("\t-m <port>   port to use for MOUNT service\n");
 		printf
 		    ("\t-t          TCP only, do not listen on UDP ports\n");
+		printf("\t-4          open sockets only with IPv4\n");
```
Asi bych to napsal jako "Use IPv4-only sockets". Nebo možná by bylo lepší "Listen only on IPv4" (nevím, co přesně to má dělat).

wentasah commented on daemon.c:
```diff
@@ -72,6 +72,7 @@ char *opt_exports = "/etc/exports";
 int opt_cluster = FALSE;
 char *opt_cluster_path = "/";
 int opt_tcponly = FALSE;
+int opt_ipv4only = FALSE;
```
Do commit message napsat, k čemu je to dobré - proč je potřeba tuhle volbu přidávat.

wentasah commented on daemon.c:
```diff
@@ -795,21 +799,22 @@ static void register_mount_service(SVCXPRT * udptransp, SVCXPRT * tcptransp)
 			"unable to register (MOUNTPROG, MOUNTVERS1, tcp6).");
 		daemon_exit(0);
 	    }
+	    /* Register MOUNT service (v3) for TCP6 */
```
Tenhle comentář dejte do commitu, který přidává kód níže.

wentasah commented on unfsd.8:
```diff
@@ -110,6 +110,14 @@ provides its services to clients using either UDP or TCP as communications
+.B \-4
+IPv4 only operation. By default,
+.B unfsd
+rovides its services to clients using either IPv4 or IPv6. When this option
```
provides

wentasah commented on daemon.c:
```diff
@@ -112,19 +112,46 @@ void logmsg(int prio, const char *fmt, ...)
- * return remote address from svc_req structure
+ * return remote address from svc_req structure in IPv6 format
+ * 
+ * IPv4 addresses are mapped to IPv6
```
IPv4 addresses are returned in so-called IPv4-mapped IPv6 address format.

wentasah commented on daemon.c:
```diff
@@ -112,19 +112,46 @@ void logmsg(int prio, const char *fmt, ...)
+ * 
+ * IPv4 addresses are mapped to IPv6
+ *
+ * for AF_INET6, remote_buf is not used and ptr inside svc_req struct is returned
+ * for AF_INET, remote_buf is used and its poineter is returned
```
Pokud je komentář víc, než jedna věta, tak věty začínejte velkým písmenem a na konci dělejte tečku. A napište to jako:

For AF_INET6 addresses, remote_buf is not used and a pointer to a field of svc_req is returned.
For AF_INET addresses, the address is stored into the remote_buf and its address is returned.

skoudmar replied:
Zde bych to změnil na:

For AF_INET6 addresses, remote_buf is not used and a pointer to a field of netbuf (provided by svc_req) is returned.

wentasah replied:
OK

wentasah commented on daemon.c:
```diff
@@ -112,19 +112,46 @@ void logmsg(int prio, const char *fmt, ...)
+ * for AF_INET, remote_buf is used and its poineter is returned
  */
-const struct in6_addr *get_remote(struct svc_req *rqstp)
+const struct in6_addr *get_remote(struct svc_req *rqstp, struct in6_addr *remote_buf) 
```
Myslím, že tohle mělo být spíš static než const. Kdyžtak na to udělejte samostatný commit.

skoudmar replied:
Const je dobře - nechceme, aby někdo do adresy zapisoval, pokud není z remote_buf.

Static být nemůže, protože je využívaná v dalších souborech.

wentasah reviewed (commented):
Posílám slíbené komentáře.

skoudmar reviewed (commented):
Provedl jsem aktualizaci na základě vašich připomínek.

CendioOssman committed:
Avoid legacy service routines

Use the current service creation and registering routines instead of
relying on the legacy support.

CendioOssman committed:
Also register IPv6 services with rpcbind

Based on patch by Martin Škoudlil <skoudmar@fel.cvut.cz>.

CendioOssman committed:
Merge branch 'tirpc'

CendioOssman committed:
Fix dependency order for yacc/bison output

It outputs both of these files, but make will treat things as two
independent rules, running two copies of yacc/bison at the same time.

CendioOssman committed:
Proper dependencies on top level

Makes sure things are built in the correct order when doing things in
parallel.

CendioOssman committed:
Remove prerequisites from install targets

Things should never be built as a side effect of installing things.

skoudmar committed:
svc_dg_create and svc_vc_create requires bound socket

When started with '-u' the socket passed to these functions is not bound.
This will prevent any connection.
